### PR TITLE
GLT-2160 Add Partial Box Search

### PR DESCRIPTION
**Pull request by Kashif-S**

### CODE REVIEW

**Comment by Kashif-S** [comment]
It might be a good idea to either add a box name column to the table or to add the box name in brackets next to the box alias, as users cannot currently see the name, which could create difficulties if a user partial searches for a box by its box name.

**Comment by djcooke** [comment]
I can understand wanting to search by a partial alias, but I'm not sure that searching by partial name or barcode is useful or desired (e.g. it makes sense to scan a barcode, but not really to type in half of it). I realize that I made the ticket and did not specify this, but we should probably ask Ilinca.

Also @Kashif-S, could you please include the JIRA ticket number in the commit message as well as the branch name? Thanks!

File: sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateBoxDao.java [hunk 1]
```diff
@@ -182,6 +184,61 @@ public BoxUse getUseById(long id) throws IOException {
...
+      throw new NullPointerException("No search String provided");
+    }
+    Criteria criteria = currentSession().createCriteria(BoxImpl.class);
+    criteria.add(Restrictions.or(
+        Restrictions.like("identificationBarcode", search, MatchMode.ANYWHERE),
+        Restrictions.like("name", search, MatchMode.ANYWHERE),
+        Restrictions.like(FIELD_ALIAS, search, MatchMode.ANYWHERE)));
+    @SuppressWarnings("unchecked")
+    List<Box> results = criteria.list();
+    results.sort(new Comparator<Box>() {
```
Comment (sqwci) [inline sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateBoxDao.java]: MAJOR Make this anonymous inner class a lambda [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1604)

File: sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateBoxDao.java [hunk 2]
```diff
@@ -182,6 +184,61 @@ public BoxUse getUseById(long id) throws IOException {
     return results;
   }
 
+  @Override
+  public List<Box> getByPartialSearch(String search) {
```
Comment (sqwci) [inline sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateBoxDao.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 24 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)

**Comment by sqwci** [comment]
SonarQube analysis reported 15 issues
* ![CRITICAL][CRITICAL] 5 critical
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultBoxService.java#L191](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultBoxService.java#L191): Refactor this method to reduce its Cognitive Complexity from 31 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [BoxRestController.java#L491](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/BoxRestController.java#L491): Refactor this method to reduce its Cognitive Complexity from 22 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [HibernateBoxDao.java#L48](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateBoxDao.java#L48): Define a constant instead of duplicating this literal "identificationBarcode" 6 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [HibernateBoxDao.java#L174](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateBoxDao.java#L174): Define a constant instead of duplicating this literal "No search String provided" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![MAJOR][MAJOR] [BoxRestController.java#L548](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/BoxRestController.java#L548): Extract this nested try block into a separate method. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)
1. ![MAJOR][MAJOR] [hot_boxable.js#L73](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/miso-web/src/main/webapp/scripts/hot_boxable.js#L73): Rename "row" as this name is already used in declaration at line 57. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)
1. ![MAJOR][MAJOR] [hot_boxable.js#L74](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/miso-web/src/main/webapp/scripts/hot_boxable.js#L74): Rename "boxAlias" as this name is already used in declaration at line 58. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)
1. ![MAJOR][MAJOR] [hot_boxable.js#L78](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/miso-web/src/main/webapp/scripts/hot_boxable.js#L78): Rename "pos" as this name is already used in declaration at line 61. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)
1. ![MINOR][MINOR] [DefaultBoxService.java#L204](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultBoxService.java#L204): Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)
1. ![MINOR][MINOR] [BoxRestController.java#L360](https://github.com/TGAC/miso-lims/blob/6392be596866bf630369e561ae4dc17a0160ce38/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/BoxRestController.java#L360): Combine this catch with the one at line 358, which has the same body. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2147)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sqwci** [comment]
SonarQube analysis reported 13 issues
* ![CRITICAL][CRITICAL] 4 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 5 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultBoxService.java#L191](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultBoxService.java#L191): Refactor this method to reduce its Cognitive Complexity from 31 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [BoxRestController.java#L491](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/BoxRestController.java#L491): Refactor this method to reduce its Cognitive Complexity from 22 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [HibernateBoxDao.java#L48](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateBoxDao.java#L48): Define a constant instead of duplicating this literal "identificationBarcode" 6 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [HibernateBoxDao.java#L174](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateBoxDao.java#L174): Define a constant instead of duplicating this literal "No search String provided" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![MAJOR][MAJOR] [BoxRestController.java#L548](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/BoxRestController.java#L548): Extract this nested try block into a separate method. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)
1. ![MAJOR][MAJOR] [hot_boxable.js#L73](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/miso-web/src/main/webapp/scripts/hot_boxable.js#L73): Rename "row" as this name is already used in declaration at line 57. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)
1. ![MAJOR][MAJOR] [hot_boxable.js#L74](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/miso-web/src/main/webapp/scripts/hot_boxable.js#L74): Rename "boxAlias" as this name is already used in declaration at line 58. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)
1. ![MAJOR][MAJOR] [hot_boxable.js#L78](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/miso-web/src/main/webapp/scripts/hot_boxable.js#L78): Rename "pos" as this name is already used in declaration at line 61. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)
1. ![MINOR][MINOR] [DefaultBoxService.java#L204](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultBoxService.java#L204): Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)
1. ![MINOR][MINOR] [BoxRestController.java#L360](https://github.com/TGAC/miso-lims/blob/d4e128ec24ed6032f119d62b5af793ad3c41d634/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/BoxRestController.java#L360): Combine this catch with the one at line 358, which has the same body. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2147)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**